framework/src/meshgenerators/CoarsenBlockGenerator.C

```diff
@@ -0,0 +1,411 @@
+                 start_elem->subdomain_id(),
+                 " set to be coarsened ",
+                 _coarsening[i],
+                 " times but the max coarsening required is ",
+                 max_c);
+
+  // Determine how many times the coarsening will be used
+  int max = *std::max_element(_coarsening.begin(), _coarsening.end());
+  if (max > 0 && !mesh->is_prepared())
+    // we prepare for use to make sure the neighbors have been found
+    mesh->prepare_for_use();
+
+  auto mesh_ptr = recursive_coarsen(block_ids, mesh, _coarsening, max, /*step=*/0);
+
+  // element neighbors are not valid
+  if (max > 0)
+    mesh_ptr->set_isnt_prepared();
+
+  return mesh_ptr;
+}
+
+std::unique_ptr<MeshBase>
+CoarsenBlockGenerator::recursive_coarsen(const std::vector<subdomain_id_type> & block_ids,
+                                         std::unique_ptr<MeshBase> & mesh,
+                                         const std::vector<unsigned int> & coarsening,
+                                         unsigned int max,
+                                         unsigned int coarse_step)
+{
+  if (coarse_step == max)
+    return dynamic_pointer_cast<MeshBase>(mesh);
+
+  // Elements should know their neighbors
+  if (!mesh->is_prepared())
+    mesh->prepare_for_use();
+
+  // We wont be modifying the starting mesh for simplicity, we will make a copy and return that
+  std::unique_ptr<MeshBase> mesh_return;
+  int max_num_coarsened = -1;
+
+  const auto base_start_elem = (*mesh->sub_point_locator())(_starting_point);
+
+  // Try every node as the 'center' point of a coarsened element
+  for (const auto & start_node_index : base_start_elem->node_index_range())
+  {
+    if (_verbose)
+      _console << "Step " << coarse_step + 1 << " coarsening attempt #" << start_node_index
+               << "\nUsing node " << *base_start_elem->node_ptr(start_node_index)
+               << " as the interior node of the coarse element." << std::endl;
+
+    // Make a copy of the mesh in case the initial node choice was bad
+    auto mesh_copy = mesh->clone();
+
+    // We will only have a single starting element for now. If there are non-connected components,
+    // we will need to have a starting element-node pair in every component.
+    auto start_elem = mesh_copy->elem_ptr(base_start_elem->id());
+    mooseAssert(start_elem, "Should have a real elem pointer");
+    mooseAssert(start_elem->active(), "Starting element must be active");
+
+    auto start_node = start_elem->node_ptr(start_node_index);
+    mooseAssert(start_node, "Starting node should exist");
+
+    // This set will keep track of all the 'fine elem' + 'coarse element interior node' pairs
+    // we should attempt to form coarse element from
+    // TODO: think about the implications of set vs vector. Set might grow to the entire mesh
+    // due to sorting. Vector we could insert at the beginning and treat new candidates immediately
+    std::set<std::pair<Elem *, Node *>> candidate_pairs;
```

Comment:
because this will sort pointers, it wont take the same path in parallel I think, hence why this cannot run in parallel even replicated

If I added a custom comparator for ordering of the set it would fix that. I also dont think we need to run this in parallel

Reply:
You can add a custom comparator

Reply:
You can look at `AutomaticMortarGeneration` which leverages `CompareDofObjectsByID` in order to get consistent sorting in parallel. You would have to implement it for the pair, but it's the same idea

Reply:
Added sorting along a direction.

Reply:
Direction will be nicer than ID ordering as in the future I can see users specify it to help the algorithm coarsen. It works fine with a random one for now

Reply:
Worked with even more sets of pointers for 3D and running into these ordering issues again.

I dont really set the need for replicated parallel mesh generation. Especially for such an expensive step where it make sense to serialize or distribute but not really replicate. 

I would be down to move forward with serial here.

Reply:
It's convenient to be able to run a single input for mesh generation and problem solve. I think that's probably the main argument for parallel replicated mesh generation

Reply:
yeah it is. I gave it another shot.

Turns out custom comparators are tricky. You have to return a boolean, but it wont complain if you return a Real.
Should work now